[lldb/Interpreter] Make ScriptedInterface Object creation more generic

medismailben opened the pull request:
This patch changes the way plugin objects used with Scripted Interfaces
are created.

Instead of implementing a different SWIG method to create the object for
every scripted interface, this patch makes the creation more generic by
re-using some of the ScriptedPythonInterface templated Dispatch code.

This patch also improves error handling of the object creation by
returning an `llvm::Expected`.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

JDevlieghere commented on lldb/include/lldb/Interpreter/Interfaces/ScriptedPlatformInterface.h:
```diff
@@ -19,11 +19,11 @@
-  StructuredData::GenericSP
+  virtual llvm::Expected<StructuredData::GenericSP>
   CreatePluginObject(llvm::StringRef class_name, ExecutionContext &exe_ctx,
                      StructuredData::DictionarySP args_sp,
-                     StructuredData::Generic *script_obj = nullptr) override {
-    return {};
+                     StructuredData::Generic *script_obj = nullptr) {
+    llvm_unreachable("unimplemented!");
```
Could this be a pure virtual method? If not maybe something like "Cannot create an instance of an interface" would be a more informative error message.

medismailben replied:
This can't be virtual pure because the `ScriptInterpreter` base class implements `virtual lldb::Scripted{Process,Platform,Thread}Interface()` methods that instantiate each base interface.

JDevlieghere commented on lldb/source/Plugins/Process/scripted/ScriptedThread.cpp:
```diff
@@ -56,14 +56,17 @@ ScriptedThread::Create(ScriptedProcess &process,
+      script_object);
 
-  if (!owned_script_object_sp)
+  if (!obj_or_err) {
     return llvm::createStringError(llvm::inconvertibleErrorCode(),
                                    "Failed to create script object.");
+  }
```
No braces around single-line if.

JDevlieghere commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.h:
```diff
@@ -32,6 +32,84 @@ class ScriptedPythonInterface : virtual public ScriptedInterface {
+    using Locker = ScriptInterpreterPythonImpl::Locker;
+
+    std::string error_string;
+    if (class_name.empty() &&
+        llvm::StringRef(m_interpreter.GetDictionaryName()).empty() &&
+        !script_obj)
+      return llvm::createStringError(
+          llvm::inconvertibleErrorCode(),
+          "ScriptedPythonInterface::CreatePluginObject - missing script class "
+          "name, dictionary or object.");
```
Can we make these 3 separate errors? We know which ones are missing. Alternatively, you could make them bools and conditionally include the "name", "dictionary" and "object" substrings based on that.

JDevlieghere commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.h:
```diff
@@ -32,6 +32,84 @@ class ScriptedPythonInterface : virtual public ScriptedInterface {
+      if (llvm::Error e = expected_return_object.takeError())
+        return e;
+      result = std::move(expected_return_object.get());
+    } else
+      result = PythonObject(PyRefType::Borrowed,
+                            static_cast<PyObject *>(script_obj->GetValue()));
```
This is mostly personal preference, but since this is simpler than the other case, I'd put this one first. That way I can page out the "what if script_obj is not NULL" as I'm reading the bigger block above.

bulbazord replied:
+1

JDevlieghere commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.h:
```diff
@@ -32,6 +32,84 @@ class ScriptedPythonInterface : virtual public ScriptedInterface {
+              class_name, dict);
+
+      if (!pfunc.IsAllocated()) {
+        error_string.append("Could not find script class: ");
+        error_string.append(class_name);
+        return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                       error_string);
```
Could you just pass those as twines to `createStringError`?

bulbazord commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.h:
```diff
@@ -32,6 +32,84 @@ class ScriptedPythonInterface : virtual public ScriptedInterface {
+    PythonObject result = {};
+
+    if (!script_obj) {
+      auto dict =
+          PythonModule::MainModule().ResolveName<python::PythonDictionary>(
+              m_interpreter.GetDictionaryName());
```
Do we need to check the validity of the dictionary?

bulbazord commented on lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedPythonInterface.h:
```diff
@@ -32,6 +32,84 @@ class ScriptedPythonInterface : virtual public ScriptedInterface {
+      result = PythonObject(PyRefType::Borrowed,
+                            static_cast<PyObject *>(script_obj->GetValue()));
+
+    if (!result.IsValid())
+      return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                     "python object result is invalid.");
```
You could augment this error to say something like "The resulting object from the call to $NAME is not a valid Python Object" or something to this effect

medismailben replied:
Thanks for the suggestions however I can't get the $name of the called passed a `script_obj`. I'll still rephrase it though.

bulbazord reviewed (commented):
Works for me generally. I'm curious to know if some of these can be pure virtual function as Jonas mentioned.

medismailben committed:
[lldb/Interpreter] Make ScriptedInterface Object creation more generic

This patch changes the way plugin objects used with Scripted Interfaces
are created.

Instead of implementing a different SWIG method to create the object for
every scripted interface, this patch makes the creation more generic by
re-using some of the ScriptedPythonInterface templated Dispatch code.

This patch also improves error handling of the object creation by
returning an `llvm::Expected`.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

JDevlieghere reviewed (approved):
LGTM

DavidSpickett commented:
We still have a failure on Windows after the follow ups:
```
# .---command stderr------------
# | Cannot create an instance of the ScriptedProcessInterface!
# | UNREACHABLE executed at C:\Users\tcwg\david.spickett\llvm-project\lldb\include\lldb/Interpreter/Interfaces/ScriptedProcessInterface.h:29!
# | PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
# | Stack dump:
# | 0.  Program arguments: c:\\users\\tcwg\\david.spickett\\build-llvm\\bin\\lldb-test.exe ir-memory-map C:\\Users\\tcwg\\david.spickett\\build-llvm\\tools\\lldb\\test\\Shell\\Expr\\Output\\TestIRMemoryMapWindows.test.tmp C:\\Users\\tcwg\\david.spickett\\llvm-project\\lldb\\test\\Shell\\Expr/Inputs/ir-memory-map-basic
# | 1.  HandleCommand(command = "run")
# | Exception Code: 0xC000001D
# |  #0 0x00007ff71e493e0c HandleAbort C:\Users\tcwg\david.spickett\llvm-project\llvm\lib\Support\Windows\Signals.inc:424:0
# |  #1 0x00007ffe20f30ef8 (C:\WINDOWS\SYSTEM32\ucrtbased.dll+0xa0ef8)
# |  #2 0x00007ffe20f32a50 (C:\WINDOWS\SYSTEM32\ucrtbased.dll+0xa2a50)
# |  #3 0x00007ff71e460358 llvm::llvm_unreachable_internal(char const *, char const *, unsigned int) C:\Users\tcwg\david.spickett\llvm-project\llvm\lib\Support\ErrorHandling.cpp:212:0
# |  #4 0x00007ff71e999fc4 lldb_private::ScriptedProcessInterface::CreatePluginObject(class llvm::StringRef, class lldb_private::ExecutionContext &, class std::shared_ptr<class lldb_private::StructuredData::Dictionary>, class lldb_private::StructuredData::Generic *) C:\Users\tcwg\david.spickett\llvm-project\lldb\include\lldb\Interpreter\Interfaces\ScriptedProcessInterface.h:28:0
```
I'm looking into it.

DavidSpickett commented:
Somehow it is getting to:
```
  virtual llvm::Expected<StructuredData::GenericSP>
  CreatePluginObject(llvm::StringRef class_name, ExecutionContext &exe_ctx,
                     StructuredData::DictionarySP args_sp,
                     StructuredData::Generic *script_obj = nullptr) {
    llvm_unreachable(
        "Cannot create an instance of the ScriptedProcessInterface!");
  }
```
Starting to wonder if we're relying on some undefined order of evaluation somewhere.

DavidSpickett commented:
Never mind, this could have been the result of using an incompatible Python version. I'll get a stacktrace with one I know is used on the bot.

DavidSpickett commented:
https://github.com/llvm/llvm-project/commit/02ef12dd808f19191a4d0a9f511c4e5a5dda59b5 fixes the Windows test. If the behaviour seems suspicious to you, I can check more details of the build but I don't know anything about `Scripted...` so I'll need you to tell me what to look for.
